server/src/main/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenServices.java

```diff
@@ -503,22 +499,22 @@ private KeyInfo getActiveKeyInfo() {
         if (null != grantType) {
             claims.put(GRANT_TYPE, grantType);
         }
-        if (user!=null && userId!=null) {
-            claims.put(USER_ID, userId);
+        if (user!=null) {
```

Comment:
I don't think this user reference can be null by the same logic above (I realize these null checks were already here).

Am I missing something?

Reply:
User can be null if client credential

Reply:
I see, I missed the condition in `createAccessToken` checking for the client credentials grant... makes sense.

But I definitely smell a refactor opportunity here. So much of this code isn't applicable to the client credentials grant (find a user, generate an id token, generate a refresh token), it's easy to make mistakes (or perform extra database lookups).

Reply:
to me this coding is fine now because the client credential flow is currently the only without a user in it, but there are other flows , e.g. saml-bearer for clients or the x509 (which I want PR) is also without a user but should grant an access token, so this check is clear to me.

The extra lookup in DB is something which we see a problem